libs/@local/hashql/core/src/arena/transaction.rs

```diff
@@ -0,0 +1,510 @@
+use core::ops::Index;
+
+use crate::id::{HasId, Id as _};
+/// A saved state of a `TransactionalArena` that can be restored at a later time.
+///
+/// Checkpoints allow for transactional operations by saving the state of the arena
+/// before making changes, and then restoring it if those changes need to be rolled back.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Checkpoint<T>(rpds::Vector<T>);
+
+/// A persistent arena that provides transactional semantics for storing and modifying objects.
+///
+/// `TransactionalArena` uses a persistent data structure internally which allows efficient
+/// creation of checkpoints and restoration to previous states. This makes it suitable for
+/// scenarios where you need to try a series of operations and potentially roll them back.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct TransactionalArena<T> {
+    items: rpds::Vector<T>,
+}
+
+impl<T> TransactionalArena<T> {
+    /// Creates a new, empty `TransactionalArena`.
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            items: rpds::Vector::new(),
+        }
+    }
+
+    /// Returns the next available ID for an item in this arena.
+    fn next_id(&self) -> usize {
+        self.items.len()
+    }
+
+    /// Adds an item to the arena using a builder function that receives the assigned ID.
+    ///
+    /// This is useful when the item's constructor needs to know its own ID.
+    ///
+    /// # Returns
+    ///
+    /// The ID assigned to the newly added item.
+    pub fn push_with(&mut self, item: impl FnOnce(T::Id) -> T) -> T::Id
+    where
+        T: HasId,
+    {
+        let id = T::Id::from_usize(self.next_id());
+
+        self.items.push_back_mut(item(id));
+        id
+    }
+
+    /// Adds an existing item to the arena.
+    ///
+    /// # Returns
+    ///
+    /// The ID assigned to the newly added item.
+    pub fn push(&mut self, item: T) -> T::Id
```

Comment:
You can use `push_with` instead

Reply:
removed `push` in [b466c4b](https://github.com/hashintel/hash/pull/6918/commits/b466c4b109aae5c83fdd623df7d1cfe297208a34), as it is prone to breaking an important invariant of the arena - `T::id` should return the assigned id, with `push` you can just push in random types where the id isn't the same.

I've also added an additional guard in [c828ccf](https://github.com/hashintel/hash/pull/6918/commits/c828ccfb057b40daa4793a9594c9481878f592a9)